File: src/main/java/dev/koen/aoc/day2/Two.java
package dev.koen.aoc.day2;

import dev.koen.aoc.common.IntPair;
import dev.koen.aoc.util.FileReader;

import java.nio.file.Path;
import java.util.function.Predicate;

public class Two {

    public static void main(String[] args) {
        final var numberOfValidPasswords = FileReader.readLines(Path.of("src/test/resources/input-two.txt"))
                .stream()
                .filter(hasValidPassword())
                .count();

        System.out.println(numberOfValidPasswords);
    }

    private static Predicate<String> hasValidPassword() {
        return line -> {
            final var parts = line.split(" ");

            final IntPair indices = readIndices(parts[0]);
            final char c = parts[1].charAt(0);

            return parts[2].charAt(indices.left()) == c
                    ^ parts[2].charAt(indices.right()) == c;
        };
    }

    private static IntPair readIndices(String part) {
        final var numbers = part.split("-");
        final var left = Integer.parseUnsignedInt(numbers[0]) - 1;
        final var right = Integer.parseUnsignedInt(numbers[1]) - 1;
        return new IntPair(left, right);
    }
}

File: src/main/kotlin/dev/koen/aoc/day2/Two.kt
package dev.koen.aoc.day2

import java.io.File

fun main() {
    File("src/test/resources/input-two.txt")
        .readLines()
        .map(String::toPolicy2)
        .filter(Policy2::isValid)
        .count()
        .let { amountOfValidPasswords -> println(amountOfValidPasswords) }
}

private fun String.toPolicy2(): Policy2 {
    val parts = this.split(" ")

    val indices = parts[0].toIndices()
    val c = parts[1].toCharArray()[0]
    val password = parts[2]

    return Policy2(indices, c, password)
}

private fun String.toIndices(): Pair<Int, Int> {
    val numbers = this.split("-")

    val start = numbers[0].toInt()
    val end = numbers[1].toInt()

    return Pair(start - 1, end - 1)
}

private data class Policy2(val pair: Pair<Int, Int>, val char: Char, val password: String) {
    fun isValid(): Boolean {
        return (password[pair.first] == char) xor (password[pair.second] == char)
    }
}